hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -1018,36 +1031,40 @@ public void onError(Channel channel, Throwable cause) {
 
     private class ConnectToAllClusterMembersTask implements Runnable {
 
-        private Set<Address> connectingAddresses = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        private Set<UUID> connectingAddresses = Collections.newSetFromMap(new ConcurrentHashMap<>());
 
         @Override
         public void run() {
+
             if (!client.getLifecycleService().isRunning()) {
                 return;
             }
 
             for (Member member : client.getClientClusterService().getMemberList()) {
-                Address address = member.getAddress();
-
-                if (client.getLifecycleService().isRunning() && getConnection(address) == null
-                        && connectingAddresses.add(address)) {
-                    // submit a task for this address only if there is no
-                    // another connection attempt for it
-                    executor.submit(() -> {
-                        try {
-                            if (!client.getLifecycleService().isRunning()) {
-                                return;
-                            }
-                            if (getConnection(member.getUuid()) == null) {
-                                getOrConnect(address);
-                            }
-                        } catch (Exception e) {
-                            EmptyStatement.ignore(e);
-                        } finally {
-                            connectingAddresses.remove(address);
-                        }
-                    });
+                UUID uuid = member.getUuid();
+
+                if (activeConnections.get(uuid) != null) {
+                    continue;
                 }
+
+                if (!connectingAddresses.add(uuid)) {
```

Comment:
I hate to keep an extra state `connectingAddresses ` when i already have `activeConnections `. How important an optimization is it to prevent multiple connection opening attempts when we can not avoid it 100% of the time. For code simplicity, I would chose not to add this extra state.

Reply:
The place that we can not avoid happens rarely. 
This code happens periodically and more likely to occur.